File: sui-tunnel-ts/src/protocol/apiCredits.ts
/**
 * API Credits protocol: a client prepays credits and spends them per API call
 * over a tunnel (the cost-benefit example — 2 on-chain txns regardless of call
 * count).
 *
 * The client (A) locks a prepaid balance; each call burns a fixed `costPerCall`
 * to the provider (B), tracked off-chain as a signed state update. Balances
 * always sum to the locked total, so every state is directly settleable. The
 * session is terminal once the remaining balance can't cover another call.
 */

import {
  Protocol,
  Party,
  Balances,
  ProtocolContext,
  protocolDomain,
} from "./Protocol";
import { concatBytes } from "../core/bytes";
import { u64ToBeBytes } from "../core/wire";

export interface ApiCreditsState {
  /** Client (A) remaining prepaid balance. */
  client: bigint;
  /** Provider (B) accrued earnings. */
  provider: bigint;
  /** Locked total — the invariant sum (client + provider). */
  total: bigint;
  /** API calls made so far (monotonic; folded into the encoding). */
  calls: bigint;
}

/** A single metered API call; the cost is fixed by the protocol. */
export interface ApiCreditsMove {
  kind: "call";
}

const DOMAIN = protocolDomain("api_credits.v1");

export class ApiCreditsProtocol
  implements Protocol<ApiCreditsState, ApiCreditsMove>
{
  readonly name = "api_credits.v1";

  /** @param costPerCall fixed price burned per API call. */
  constructor(private readonly costPerCall: bigint = 10n) {
    if (costPerCall <= 0n) throw new Error("costPerCall must be positive");
  }

  initialState(ctx: ProtocolContext): ApiCreditsState {
    return {
      client: ctx.initialBalances.a,
      provider: ctx.initialBalances.b,
      total: ctx.initialBalances.a + ctx.initialBalances.b,
      calls: 0n,
    };
  }

  applyMove(
    state: ApiCreditsState,
    move: ApiCreditsMove,
    by: Party,
  ): ApiCreditsState {
    if (by !== "A") throw new Error("only the client (A) makes calls");
    if (move.kind !== "call") throw new Error(`unknown move: ${move.kind}`);
    if (state.client < this.costPerCall) {
      throw new Error("out of credits: remaining balance can't cover a call");
    }
    return {
      client: state.client - this.costPerCall,
      provider: state.provider + this.costPerCall,
      total: state.total,
      calls: state.calls + 1n,
    };
  }

  encodeState(state: ApiCreditsState): Uint8Array {
    return concatBytes([
      DOMAIN,
      u64ToBeBytes(state.client),
      u64ToBeBytes(state.provider),
      u64ToBeBytes(state.calls),
    ]);
  }

  balances(state: ApiCreditsState): Balances {
    return { a: state.client, b: state.provider };
  }

  isTerminal(state: ApiCreditsState): boolean {
    return state.client < this.costPerCall; // can't afford another call
  }

  randomMove(
    state: ApiCreditsState,
    by: Party,
    _rng: () => number,
  ): ApiCreditsMove | null {
    if (by !== "A" || state.client < this.costPerCall) return null;
    return { kind: "call" };
  }
}

File: rust/protocols/api-credits/tests/api_credits.rs
use tunnel_api_credits::{ApiCredits, ApiCreditsMove};
use tunnel_harness::{Balances, Protocol, Seat, TunnelContext};

fn ctx() -> TunnelContext {
    TunnelContext {
        tunnel_id: "credits-1".into(),
        initial: Balances { a: 100, b: 0 },
        seat: Seat::A,
    }
}

#[test]
fn calls_shift_fixed_cost_from_client_to_provider() {
    let protocol = ApiCredits::new(10).unwrap();
    assert_eq!(protocol.name(), "api_credits.v1");

    let initial = protocol.initial_state(&ctx());
    assert_eq!(protocol.balances(&initial), Balances { a: 100, b: 0 });
    assert_eq!(initial.calls, 0);

    let next = protocol
        .apply_move(&initial, &ApiCreditsMove::Call, Seat::A)
        .unwrap();
    assert_eq!(next.client, 90);
    assert_eq!(next.provider, 10);
    assert_eq!(next.calls, 1);
    assert_eq!(protocol.balances(&next).sum(), 100);
}

#[test]
fn only_client_can_call_until_remaining_credit_is_too_small() {
    let protocol = ApiCredits::new(30).unwrap();
    let mut state = protocol.initial_state(&ctx());

    assert!(protocol
        .apply_move(&state, &ApiCreditsMove::Call, Seat::B)
        .is_err());

    for _ in 0..3 {
        state = protocol
            .apply_move(&state, &ApiCreditsMove::Call, Seat::A)
            .unwrap();
    }

    assert_eq!(state.client, 10);
    assert!(protocol.is_terminal(&state));
    assert!(protocol.sample_move(&state, Seat::A, &mut || 0.5).is_none());
}

#[test]
fn encode_state_is_domain_tagged_and_big_endian() {
    let protocol = ApiCredits::new(10).unwrap();
    let state = protocol.initial_state(&ctx());
    let encoded = protocol.encode_state(&state);

    assert!(encoded.starts_with(b"sui_tunnel::proto::api_credits.v1"));
    assert_eq!(
        encoded.len(),
        b"sui_tunnel::proto::api_credits.v1".len() + 24
    );
    assert_eq!(
        &encoded[encoded.len() - 24..encoded.len() - 16],
        &100u64.to_be_bytes()
    );
    assert_eq!(
        &encoded[encoded.len() - 16..encoded.len() - 8],
        &0u64.to_be_bytes()
    );
    assert_eq!(&encoded[encoded.len() - 8..], &0u64.to_be_bytes());
}
